sdk/storage/src/common/xml_wrapper.cpp

```diff
@@ -0,0 +1,391 @@
+{
+  m_reader = xmlReaderForMemory((const char*)buffer, size, NULL, 0, 0);
+}
+
+xml_text_reader_wrapper::~xml_text_reader_wrapper()
+{
+  if (m_reader != nullptr)
+  {
+    xmlFreeTextReader(m_reader);
+    m_reader = nullptr;
+  }
+}
+
+bool xml_text_reader_wrapper::read() { return xmlTextReaderRead(m_reader) == 1; }
+
+unsigned xml_text_reader_wrapper::get_node_type() { return xmlTextReaderNodeType(m_reader); }
+
+bool xml_text_reader_wrapper::is_empty_element()
+{
+  return xmlTextReaderIsEmptyElement(m_reader) == 1;
+}
+
+std::string xml_text_reader_wrapper::get_local_name()
+{
+  auto xml_char = xmlTextReaderLocalName(m_reader);
+  std::string result;
+
+  if (xml_char != nullptr)
+  {
+    result = xml_char_to_string(xml_char);
+    xmlFree(xml_char);
+  }
+
+  return result;
+}
+
+std::string xml_text_reader_wrapper::get_value()
+{
+  auto xml_char = xmlTextReaderValue(m_reader);
+  std::string result;
+
+  if (xml_char != nullptr)
+  {
+    result = xml_char_to_string(xml_char);
+    xmlFree(xml_char);
+  }
+  return result;
+}
+
+bool xml_text_reader_wrapper::move_to_first_attribute()
+{
+  return xmlTextReaderMoveToFirstAttribute(m_reader) == 1;
+}
+
+bool xml_text_reader_wrapper::move_to_next_attribute()
+{
+  return xmlTextReaderMoveToNextAttribute(m_reader) == 1;
+}
+
+xml_element_wrapper::~xml_element_wrapper() {}
+
+xml_element_wrapper::xml_element_wrapper(xmlNode* node)
+{
+  m_ele = node;
+  m_ele->_private = this;
+}
+
+xml_element_wrapper* xml_element_wrapper::add_child(
+    const std::string& name,
+    const std::string& prefix)
+{
+  xmlNs* ns = nullptr;
+  xmlNode* child = nullptr;
+
+  if (m_ele->type != XML_ELEMENT_NODE)
+  {
+    return nullptr;
+  }
+
+  if (!prefix.empty())
+  {
+    ns = xmlSearchNs(m_ele->doc, m_ele, (const xmlChar*)prefix.c_str());
+    if (!ns)
+    {
+      return nullptr;
+    }
+  }
+
+  child = xmlNewNode(ns, (const xmlChar*)name.c_str()); // mem leak?
+
+  if (!child)
+    return nullptr;
+
+  xmlNode* node = xmlAddChild(m_ele, child);
+  if (!node)
+    return nullptr;
+
+  node->_private = new xml_element_wrapper(node);
+  return reinterpret_cast<xml_element_wrapper*>(node->_private);
+}
+
+void xml_element_wrapper::set_namespace_declaration(
+    const std::string& uri,
+    const std::string& prefix)
+{
+  xmlNewNs(
+      m_ele,
+      (const xmlChar*)(uri.empty() ? nullptr : uri.c_str()),
+      (const xmlChar*)(prefix.empty() ? nullptr : prefix.c_str()));
+}
+
+void xml_element_wrapper::set_namespace(const std::string& prefix)
+{
+  xmlNs* ns
+      = xmlSearchNs(m_ele->doc, m_ele, (const xmlChar*)(prefix.empty() ? nullptr : prefix.c_str()));
+  if (ns)
+  {
+    xmlSetNs(m_ele, ns);
+  }
+}
+
+void xml_element_wrapper::set_attribute(
+    const std::string& name,
+    const std::string& value,
+    const std::string& prefix)
+{
+  xmlAttr* attr = 0;
+
+  if (prefix.empty())
+  {
+    attr = xmlSetProp(m_ele, (const xmlChar*)name.c_str(), (const xmlChar*)value.c_str());
```

Comment:
Why are we putting a bunch of do nothing C style casts everywhere? That's super dangerous.

Reply:
The XML parser is a copy-paste from the current C++ Track 1 SDK, to quickly enable storage's scenario. This will be re-written.